File: python/digi/application.py

from digi.broker import BrokerProvider
from digi.reporter import Reporter
from digi.trader import Trader
from digi.transaction_storage import TransactionStorage

class Application(object):

    def __init__(self):
        self.__brokers = [\
            BrokerProvider.provide(broker) for broker in\
                BrokerProvider.get_available_brokers_id()
                         ]
        self.__storage = TransactionStorage()
        self.__trader = Trader(self.__brokers, self.__storage)
        self.__reporter = Reporter(self.__storage)

    def process_order(self, order):
        return self.__trader.process_order(order)

    def report_brokers_volume(self):
        return self.__reporter.get_brokers_transacted_volume()

    def report_clients_net_position(self):
        return self.__reporter.get_clients_net_position()

File: cpp/digi/src/application.cpp
#include <application.hpp>
#include <broker.hpp>
#include <trader.hpp>
#include <memory>

namespace digi
{

    TradeApplication::TradeApplication():brokers_(),
                                         storage_(new Storage()),
                                         trader_(new Trader(brokers_, *storage_)),
                                         reporter_(new Reporter(*storage_))
    {
        using namespace std;
        brokers_.push_back(move(Broker::factory("Broker1", Broker::Commission::CONSTANT)));
        brokers_.push_back(move(Broker::factory("Broker2", Broker::Commission::VOLUME_BASED)));
    }

    double TradeApplication::execute_order(const Order &order)
    {
        return trader_->execute_order(order);
    }

    double TradeApplication::get_client_net_position(const std::string & client_name) const
    {
        return reporter_->get_client_net_position(client_name);
    }

    int TradeApplication::get_broker_total_volume(const std::string & broker_name) const
    {
        return reporter_->get_broker_total_volume(broker_name);
    }

    const std::set<std::string> & TradeApplication::get_broker_names() const
    {
        return storage_->get_broker_names();
    }

    const std::set<std::string> & TradeApplication::get_client_names() const
    {
        return storage_->get_client_names();
    }


}
